File: Tutor Demos/Week 6/Java/Customer.java
import java.util.LinkedList;

/**
 * A customer has a savings account and a loan account.
 *
 * You can deposit into and withdraw from the savings account,
 * and you can transfer money from the savings to the loan account.
 * You can also view the balance of all accounts.
 */
public class Customer {
    public static void main(String[] args) {
        new Customer().use();
    }
    private final LinkedList<Account> accounts;

    public Customer() {
        accounts = new LinkedList<Account>();
    }

    public void use() {
        char choice;
        while ((choice = readChoice()) != 'x') {
            switch (choice) {
            case 'a': add(); break;
            case 'r': remove(); break;
            case 'd': deposit(); break;
            case 'w': withdraw(); break;
            case 's': show(); break;
            default: help(); break;
            }
        }
        System.out.println("Done");
    }

    private char readChoice() {
        System.out.print("Customer menu (a/r/d/w/s/x): ");
        return In.nextChar();
    }

    private String readType() {
        System.out.print("Account type: ");
        return In.nextLine();
    }
    private Account account(String type) {
        for (Account a : accounts) {
            if (a.has_type(type)) {
                return a;
            }
        }
        return null;
    }

    private void add() {
        String type = readType();
        Account acc = account(type);
        if (acc == null) {
            accounts.add(new Account(type));
        }
        else {
            System.out.println("Account already exists!");
        }
    }

    private void remove() {
        String type = readType();
        Account acc = account(type);
        if (acc != null) {
            accounts.remove(acc);
        }
        else {
            System.out.println("No such account!");
        }
    }

    private void deposit() {
        String type = readType();
        Account acc = account(type);
        if (acc != null) {
            acc.deposit(readAmount("deposit"));
        }
        else {
            System.out.println("No such account!");
        }
    }

    private void withdraw() {
        String type = readType();
        Account acc = account(type);
        if (acc != null) {
            double amount = readAmount("withdraw");
            if (acc.has(amount))
                acc.withdraw(amount);
            else
                System.out.println("Insufficient funds");
        }
        else {
            System.out.println("No such account!");
        }
    }

    private double readAmount(String action) {
        System.out.print("Amount to " + action + ": $");
        return In.nextDouble();
    }

    private void show() {
        for (Account a : accounts) {
            System.out.println(a);
        }
    }

    private void help() {
        System.out.println("Menu options");
        System.out.println("a = add");
        System.out.println("r = remove");
        System.out.println("d = deposit");
        System.out.println("w = withdraw");
        System.out.println("s = show");
        System.out.println("x = exit");
    }
}

File: Tutor Demos/Week 6/Python/customer.py
from account import Account

class Customer:
    def __init__(self):
        self.accounts = []

    def use(self):
        choice = ''
        while (choice := input("Customer menu (a/r/d/w/s/x): ")) != 'x':
            if choice == 'a':
                self.add()
            elif choice =='r':
                self.remove()
            elif choice == 'd':
                self.deposit()
            elif choice == 'w':
                self.withdraw()
            elif choice == 's':
                self.show()
            else:
                self.help()
        print("Done")

    #Not strictly necessary in python, but a very reused and error prone piece of code
    #Good to have only 1 instance of it
    def read_amount(self, action):
        return float(input(f"Amount to {action}: $"))

    def account(self, type):
        for acc in self.accounts:
            if acc.is_type(type):
                return acc
        return None

    def add(self):
        type = input("Account type: ")
        acc = self.account(type)
        if acc == None:
            self.accounts.append(Account(type))
        else:
            print("Account already exists!")

    def remove(self):
        type = input("Account type: ")
        acc = self.account(type)
        if acc != None:
            self.accounts.remove(acc)
        else:
            print("No such account!")

    def deposit(self):
        type = input("Account type: ")
        acc = self.account(type)
        if acc != None:
            amount = self.read_amount("deposit")
            acc.deposit(amount)
        else:
            print("No such account!")

    def withdraw(self):
        type = input("Account type: ")
        acc = self.account(type)
        if acc != None:
            amount = self.read_amount("withdraw")
            if acc.has(amount):
                acc.withdraw(amount)
            else:
                print("Insufficient funds")
        else:
            print("No such account!")

    def show(self):
        for account in self.accounts:
            print(account.__str__())

    def help(self):
        print("Menu options")
        print("a = add")
        print("r = remove")
        print("d = deposit")
        print("w = withdraw")
        print("s = show")
        print("x = exit")

if __name__ == "__main__":
    Customer().use()